### PR TITLE
Add agent tooling: linter rule scaffold, validate:pr, and skill

### PR DESCRIPTION
**Pull request by markcowl**
## Summary

Adds three components to improve AI agent performance when creating linter rules:

### 1. `eng/scripts/create-linter-rule.ts` (scaffold script)
Generates all required files for a new linter rule:
- Rule source file with correct template
- Test file with valid/invalid test stubs
- Documentation page with correct structure
- Automatically registers the rule in \linter.ts\

Usage: `pnpm create:linter-rule <name> --package <azure-core|azure-resource-manager> --severity <warning|error>`

### 2. `eng/scripts/validate-pr.ts` (pre-PR validation)
General validation script for any code change (not linter-rule-specific):
- Branch up to date check
- Build, test, lint, format, cspell checks
- Regen-docs freshness check
- Changeset existence check
- Diff cleanliness check

Usage: `pnpm validate:pr [--fix] [--skip-build] [--skip-test]`

### 3. `.github/skills/create-linter-rule/SKILL.md`
9-step TDD workflow skill for Copilot agents creating linter rules. Covers identification, scaffolding, test-first development, implementation, ruleset registration, documentation, changeset creation, and final validation.

## Testing
- Both scripts syntax-check cleanly (`
ode --check`)
- `create-linter-rule --dry-run` produces expected output
- `validate:pr --help` shows usage

Full integration testing requires `pnpm install` (not run in this environment).

### CODE REVIEW

File: .github/skills/create-linter-rule/SKILL.md [hunk 1]
```diff
@@ -0,0 +1,203 @@
+---
+name: create-linter-rule
+description: >
+  Create a new TypeSpec linter rule with TDD approach, including implementation,
+  tests, documentation, ruleset registration, and changeset. Use this skill when
+  asked to create, add, or implement a new linter rule for TypeSpec Azure libraries.
```
Comment (markcowl) [inline .github/skills/create-linter-rule/SKILL.md]: Make sure there are enough keywords in the description to make the skill fire

File: .github/skills/create-linter-rule/SKILL.md [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+- `website/src/content/docs/docs/libraries/<pkg>/rules/<rule-name>.md` — docs skeleton
+- Updates `packages/<pkg>/src/linter.ts` — registers the rule
+
+## Step 3: WRITE FAILING TESTS FIRST (TDD)
+
+Edit `packages/<pkg>/test/rules/<rule-name>.test.ts`:
+
+1. Write tests for **valid code** that should produce no diagnostics (`.toBeValid()`)
+2. Write tests for **invalid code** that should produce specific diagnostics (`.toEmitDiagnostics()`)
+3. Write tests for **edge cases**:
```
Comment (markcowl) [inline .github/skills/create-linter-rule/SKILL.md]: Here I would say to create equivalence classes for the handling of the expected input and write tests so that at least one instance of each equivalence class is included in the test.  For example, if a rule applies to ModelProperty types, equivalence classes might include: (1) simply defined properties, (2) properties defined using spread or is, (3) properties inherited from a base class.  In addition, the particulars of the linting rule will help inform the equivalence classes and boundary conditions that should be tested:  a rule about ModelProperty name prefixes would include equivalence classes like 'properties with the name prefix'.  'properties with the prefix text included at the middle or end', properties with a name shorter than the prefix, etc.

File: .github/skills/create-linter-rule/SKILL.md [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+- `packages/<pkg>/src/rules/<rule-name>.ts` — rule skeleton
+- `packages/<pkg>/test/rules/<rule-name>.test.ts` — test skeleton
+- `website/src/content/docs/docs/libraries/<pkg>/rules/<rule-name>.md` — docs skeleton
+- Updates `packages/<pkg>/src/linter.ts` — registers the rule
+
+## Step 3: WRITE FAILING TESTS FIRST (TDD)
+
+Edit `packages/<pkg>/test/rules/<rule-name>.test.ts`:
+
+1. Write tests for **valid code** that should produce no diagnostics (`.toBeValid()`)
```
Comment (markcowl) [inline .github/skills/create-linter-rule/SKILL.md:56]: Here, you should always include at least one test that ensures that library types in azure.core and azure.resourcemanager are not subject to the rule.

File: .github/skills/create-linter-rule/SKILL.md [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+
+All tests should pass. If not, fix the implementation until they do.
+
+## Step 6: REGISTER IN RULESETS
+
+Add the rule to the appropriate ruleset(s):
+
+- **Data-plane rules**: Edit `packages/typespec-azure-rulesets/src/rulesets/data-plane.ts`
+- **ARM rules**: Edit `packages/typespec-azure-rulesets/src/rulesets/resource-manager.ts`
+
```
Comment (markcowl) [inline .github/skills/create-linter-rule/SKILL.md:130]: Need to make sure it is clear that TCGC rules generally go in *both rulesets*, as do rules implemented in typespec-azure-core that are intended to apply to both ARM and data-plane specs.

File: .github/skills/create-linter-rule/SKILL.md [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+New linter rules MUST NOT break existing Azure service specs. After pushing your PR:
+
+1. Apply the `int:azure-specs` label to the PR to trigger the External Integration check
+2. This workflow packages your changes and runs TypeSpec validation against all specs in `Azure/azure-rest-api-specs`
+3. Wait for the check to pass before requesting review
+
+If the check fails, your rule produces diagnostics on existing specs. To resolve:
+
+1. **Apply an API-neutral fix to the spec** (preferred): If the fix doesn't change API behavior, submit a PR to `Azure/azure-rest-api-specs` on the **main** branch.
```
Comment (markcowl) [inline .github/skills/create-linter-rule/SKILL.md:196]: Can skills in this repo actually do this? Please verify.

File: .github/skills/create-linter-rule/SKILL.md [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+
+This checks: branch is up to date, build passes, tests pass, lint passes, format is clean, spelling is clean, regen-docs is clean, changeset exists, and diff only contains expected files.
+
+If any check fails, fix the issue and re-run. Use `pnpm validate:pr --fix` to auto-fix formatting and lint issues.
+
+## Step 10: EXTERNAL INTEGRATION CHECK
+
+New linter rules MUST NOT break existing Azure service specs. After pushing your PR:
+
+1. Apply the `int:azure-specs` label to the PR to trigger the External Integration check
```
Comment (markcowl) [inline .github/skills/create-linter-rule/SKILL.md]: Can agents actualkly do this?  If not, consider telling the agent about the manual steps to reproduce the issue.

File: eng/scripts/create-linter-rule.ts [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+  packageDir: string;
+  docsLibraryDir: string;
+  packageNpmName: string;
+  testHostComment?: string;
+  testHostImport: string;
+  testHostFactory: string;
+}
+
+const usage =
+  "Usage: pnpm create:linter-rule <rule-name> [--package <azure-core|azure-resource-manager|client-generator-core>] [--severity <warning|error>] [--description <text>] [--dry-run]";
```
Comment (markcowl) [inline eng/scripts/create-linter-rule.ts]: We should not include severity as an option at this time.  All linter rules are warnings

File: website/src/content/docs/docs/howtos/contributing/when-to-write-a-linting-rule.md [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+should exist.
+
+## How Linter Diagnostics Work
+
+All linter rules emit diagnostics with **warning** severity. This is important to understand:
+
+- **Warnings** are suppressible — authors can add a suppression comment with justification
+- **Unsuppressed warnings block CI** — the compiler treats unsuppressed diagnostics as errors
+  during validation
+- **Errors** (non-linter) indicate inconsistent or uncompilable code and are _never_ suppressible
```
Comment (markcowl) [inline website/src/content/docs/docs/howtos/contributing/when-to-write-a-linting-rule.md]:
```suggestion
- **Errors** (non-linter) indicate inconsistent code or code that will not compile and are _never_ suppressible
```

File: website/src/content/docs/docs/howtos/contributing/when-to-write-a-linting-rule.md [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+1. **Can violations be detected deterministically with high fidelity?** (detection axis)
+2. **Can violations be resolved without contextual judgment?** (resolution axis)
+
+These combine into three categories:
+
+| Detection                          | Resolution                 | Approach                                                |
+| ---------------------------------- | -------------------------- | ------------------------------------------------------- |
+| Deterministic, low false positives | Unambiguous or few options | **Standalone linting rule** — detects and may offer fix |
+| Deterministic, low false positives | Requires judgment/context  | **Hybrid** — rule detects, AI assists resolution        |
+| Requires semantic judgment         | Requires judgment/context  | **Defer to AI tooling**                                 |
```
Comment (markcowl) [inline website/src/content/docs/docs/howtos/contributing/when-to-write-a-linting-rule.md:110]: There is logically a fourth entry in this table, 'Requires Semantic Judgment' for Detection and 'Unambiguous or few options' for resolution  this should also be deferred to ai tooling

File: website/src/content/docs/docs/howtos/contributing/when-to-write-a-linting-rule.md [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+
+### ❌ Skipping a rule because the fix is hard
+
+If detection is reliable, create the rule even if the fix requires judgment. The hybrid pattern
+exists for exactly this case. A diagnostic that says "this model uses `Record<unknown>`, which
+limits SDK usability" is valuable even without an auto-fix.
+
+### ❌ Rules whose easiest fix is meaningless mechanical compliance
+
+If the rule incentivizes authors to add useless placeholders (e.g., `@doc("The Foo property")`)
```
Comment (markcowl) [inline website/src/content/docs/docs/howtos/contributing/when-to-write-a-linting-rule.md:209]: Need to add the word 'incentivizes' to the cspell dictionary

File: website/src/content/docs/docs/howtos/contributing/creating-linter-rules.md [hunk 11]
```diff
@@ -0,0 +1,596 @@
...
+That prevents follow-up churn across `linter.ts`, tests, and rulesets.
+:::
+
+### Step 2: Scaffold the Files
+
+Use the repo scaffold command:
+
+<!-- prettier-ignore -->
+```bash
+pnpm create:linter-rule <rule-name> --package <azure-core|azure-resource-manager|client-generator-core> --severity <warning|error> --description "What the rule enforces"
```
Comment (markcowl) [inline website/src/content/docs/docs/howtos/contributing/creating-linter-rules.md]: In general, please apply changes from elsewhere in this pr to this document as appropriate.  Here, we should *not* have a severity setting in the create:linter-rule command

File: website/src/content/docs/docs/howtos/contributing/creating-linter-rules.md [hunk 12]
```diff
@@ -0,0 +1,596 @@
...
+    union Color {
+      string,
+      red: "red",
+    }
+  `);
+```
+
+What to test:
+
+- Valid code that must stay silent
```
Comment (markcowl) [inline website/src/content/docs/docs/howtos/contributing/creating-linter-rules.md:212]: Mke changes I suggested in the skill definition

File: .github/skills/create-linter-rule/SKILL.md [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+- Decorators from `typespec-client-generator-core`, or rules only about client SDK generation → `typespec-client-generator-core`
+- Rules specific to ARM APIs → `typespec-azure-resource-manager`
+- Rules that apply to both data-plane and ARM, or only data-plane → `typespec-azure-core`
+
+## Step 2: SCAFFOLD
+
+Generate all required files using the repo's scaffolding tool:
+
+```bash
+pnpm create:linter-rule < rule-name > --package < azure-core | azure-resource-manager | client-generator-core > --description "<description>"
```
Comment (timotheeguerin) [inline .github/skills/create-linter-rule/SKILL.md:42]: just wondering if maintaining this scaffolding script is worth it? Would the AI not just be able to pattern match if we list the files to create?

File: .github/skills/create-linter-rule/SKILL.md [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+
+```bash
+pnpm --filter "@azure-tools/typespec-<pkg>..." build
+pnpm --filter "@azure-tools/typespec-<pkg>" regen-docs
+```
+
+## Step 8: CREATE CHANGESET
+
+```bash
+pnpm change add
```
Comment (timotheeguerin) [inline .github/skills/create-linter-rule/SKILL.md:163]: can also use `<packges...> --message "abc"` and `--kind` to set the values, might work better than interactive cli

File: .github/skills/create-linter-rule/SKILL.md [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+
+This checks: branch is up to date, build passes, tests pass, lint passes, format is clean, spelling is clean, regen-docs is clean, changeset exists, and diff only contains expected files.
+
+If any check fails, fix the issue and re-run. Use `pnpm validate:pr --fix` to auto-fix formatting and lint issues.
+
+## Step 10: EXTERNAL INTEGRATION CHECK
+
+New linter rules MUST NOT break existing Azure service specs. After pushing your PR:
+
+1. Apply the `int:azure-specs` label to the PR to trigger the External Integration check: `gh pr edit --add-label "int:azure-specs"`
```
Comment (timotheeguerin) [inline .github/skills/create-linter-rule/SKILL.md:188]: feel like we should add a policy to add this label automatically when rules are modified

File: .github/skills/create-linter-rule/SKILL.md (hunk 13 above)
Comment (markcowl) [reply .github/skills/create-linter-rule/SKILL.md:42]: It's a good question,  the idea is that it makes it more likely that even without the skill, agents will pick up the scaffolding (which is why the validation script as well).

Whenever we add new collateral to linting rules, we would have to remember to update, or we could make this part of the default agent instructions, perhaps?

File: .github/skills/create-linter-rule/SKILL.md (hunk 15 above)
Comment (markcowl) [reply .github/skills/create-linter-rule/SKILL.md:188]: Agreed.  Can we add this in a separate PR?  Or do you think we should just remove this?

File: .github/skills/create-linter-rule/SKILL.md (hunk 14 above)
Comment (markcowl) [reply .github/skills/create-linter-rule/SKILL.md:163]: @copilot please update the skill to include the fact that the change text and change kind can be added directly on the command line

File: .github/skills/create-linter-rule/SKILL.md (hunk 13 above)
Comment (timotheeguerin) [reply .github/skills/create-linter-rule/SKILL.md:42]: yeah the more I think about it, this feels like a good idea(save on token for the part that is consitent)

**Review by timotheeguerin — APPROVED** [review]

File: .github/skills/create-linter-rule/SKILL.md [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+This creates:
+
+- `packages/<pkg>/src/rules/<rule-name>.ts` — rule skeleton
+- `packages/<pkg>/test/rules/<rule-name>.test.ts` — test skeleton
+- `website/src/content/docs/docs/libraries/<pkg>/rules/<rule-name>.md` — docs skeleton
+- Updates `packages/<pkg>/src/linter.ts` — registers the rule
+
+## Step 3: WRITE FAILING TESTS FIRST (TDD)
+
+Edit `packages/<pkg>/test/rules/<rule-name>.test.ts`:
```
Comment (iscai-msft) [inline .github/skills/create-linter-rule/SKILL.md:54]: can we add validation / naming conventions to the skill as well? i.e. what @timotheeguerin was saying here

```
Current proposal is
- Use kebab-case rule IDs.
- DO NOT include the package name again in the code
- Use no-<thing> when the rule bans a construct or usage.
- Use use-<preferred-thing> when the rule points users to a standard/preferred TypeSpec pattern.
- Keep name short and concise as much as possible
```

File: website/src/content/docs/docs/howtos/contributing/when-to-write-a-linting-rule.md [hunk 17]
```diff
@@ -0,0 +1,215 @@
+---
+title: When to Write a Linting Rule vs. Defer to AI Tooling
+---
+
+This guidance helps authors decide whether a given Azure API design guideline should be
```
Comment (iscai-msft) [inline website/src/content/docs/docs/howtos/contributing/when-to-write-a-linting-rule.md]: can you also make clear what the diff between errors, linting rules, and best practice linting rules are?

**Review by iscai-msft — APPROVED** [review]
amazing thank you so much @markcowl for the quick turnaround!
